webpack/components/Table/TableWrapper.js

```diff
@@ -0,0 +1,108 @@
+import React, { useEffect, useState } from 'react';
+import {
+  Table,
+  TableHeader,
+  TableBody,
+} from '@patternfly/react-table';
+import {
+  Pagination,
+} from '@patternfly/react-core';
+import PropTypes from 'prop-types';
+import { useDispatch } from 'react-redux';
+import { STATUS } from 'foremanReact/constants';
+import { useForemanSettings } from 'foremanReact/Root/Context/ForemanContext';
+
+import EmptyStateMessage from './EmptyStateMessage';
+import Loading from './Loading';
+
+/* Patternfly 4 table wrapper */
+const TableWrapper = ({
+  status, cells, rows, error, emptyTitle, emptyBody, metadata, fetchItems, ...extraTableProps
+}) => {
+  const dispatch = useDispatch();
+  const { foremanPerPage = 20 } = useForemanSettings();
+  // setting pagination to local state so it doesn't disappear when page reloads
```

Comment:
Seems like two sources of truth, `metadata` and `useForemanSettings()`.  Is there a better way?

Looking in `ContentViewsTable.js`, I see `metadata` is part of `response`, which Foreman is sending us anyway.  Isn't Foreman sending us the same pagination settings in the API response that we get thru React context and `useForemanSettings`?  If the data is the same, can this be simplified to use a single source of truth?

Reply:
There is a default per page setting in Foreman. You can play around with this by going to `Settings > General` tab and changing `Entries per page` to 30 or something like that and you will see 30 load at first. This actually isn't working correctly in the angular Katello pages from what I can tell :/ but it is working on the Subscriptions page so it seems to be accounted for in the pf3 pages.

I'm updating this so the logic is:
- Get per page from Foreman settings first with 20 as a fallback and use 0 as a default for total and 1 for page/
- Update page and perPage, and total (though total shouldn't change here) when pagination is updated
- Updating page, perPage, and total when metadata comes back in API response. The API sends back the page and per-page of what was sent so this will stay the same, but we need the total to be set by the response.

I have this a separate function inside the TableWrapper right now and will push up soon. I'm not sure I see the use case of making it a separate hook, the data is pretty specific to the pf4 table and it looks like the pf3 pagination is structured differently. Is there a particular component you have in mind?

I also updated to use the pagination options from Foreman, so when you set a default pagination setting, it is now an option. This is the same as how the pf3 table works. For example, 30 is the default per page setting and is now listed as an option in the dropdown:
![h](http://www.clipular.com/c/5583786279174144.png?k=kmbigfe7FfgiC47D-ZJw3wY5nJ0)

Reply:
Thanks for the explanation.  Makes sense not to have a separate hook if there's nowhere else you would use it.